File: sdks/swift/Sources/SequinsOtel/Sequins.swift
import Foundation
import Logging
import Metrics
import OTel
import OTLPHTTPExporter

/// Configuration for the Sequins OpenTelemetry SDK.
public struct SequinsConfig {
    /// The name of your service. Defaults to `OTEL_SERVICE_NAME` env var, or `"unknown_service"`.
    public var serviceName: String?
    /// The OTLP HTTP endpoint. Defaults to `OTEL_EXPORTER_OTLP_ENDPOINT` or `"http://localhost:4318"`.
    public var endpoint: String?

    public init(serviceName: String? = nil, endpoint: String? = nil) {
        self.serviceName = serviceName
        self.endpoint = endpoint
    }
}

/// Handle returned by `Sequins.initialize(...)`. Provides access to all configured OTel providers.
public final class SequinsHandle {
    /// The configured tracer provider. Use `tracerProvider.tracer(name:)` for manual instrumentation.
    public let tracerProvider: OTelTracerProvider
    /// Call this before your application exits to flush pending telemetry.
    public let shutdown: () -> Void

    init(tracerProvider: OTelTracerProvider, shutdown: @escaping () -> Void) {
        self.tracerProvider = tracerProvider
        self.shutdown = shutdown
    }
}

/// Sequins OpenTelemetry SDK — zero-config observability for local development.
///
/// Configures OpenTelemetry traces to export to a local Sequins instance via OTLP/HTTP.
///
/// ```swift
/// import SequinsOtel
///
/// let sequins = try Sequins.initialize(serviceName: "my-app")
/// let tracer = sequins.tracerProvider.tracer(name: "my-module")
///
/// // On exit:
/// sequins.shutdown()
/// ```
public enum Sequins {
    private static let defaultEndpoint = "http://localhost:4318"

    /// Initializes OpenTelemetry pre-configured for Sequins.
    ///
    /// - Parameter serviceName: The name of your service.
    /// - Parameter endpoint: The OTLP HTTP endpoint. Defaults to `OTEL_EXPORTER_OTLP_ENDPOINT` or `http://localhost:4318`.
    /// - Returns: A `SequinsHandle` with the configured provider and shutdown function.
    public static func initialize(
        serviceName: String? = nil,
        config: SequinsConfig = SequinsConfig()
    ) throws -> SequinsHandle {
        let resolvedServiceName = config.serviceName
            ?? serviceName
            ?? ProcessInfo.processInfo.environment["OTEL_SERVICE_NAME"]
            ?? "unknown_service"

        let resolvedEndpoint = config.endpoint
            ?? ProcessInfo.processInfo.environment["OTEL_EXPORTER_OTLP_ENDPOINT"]
            ?? defaultEndpoint

        guard let endpointURL = URL(string: resolvedEndpoint) else {
            throw SequinsError.invalidEndpoint(resolvedEndpoint)
        }

        let resource = OTelResource(attributes: [
            "service.name": .string(resolvedServiceName),
        ])

        let exporter = try OTLPHTTPSpanExporter(
            endpoint: endpointURL.appendingPathComponent("v1/traces"),
            headers: ["Content-Type": "application/x-protobuf"]
        )

        let processor = OTelBatchSpanProcessor(
            exporter: exporter,
            configuration: .init()
        )

        let tracerProvider = OTelTracerProvider(
            resource: resource,
            sampler: OTelConstantSampler(isOn: true),
            processor: processor
        )

        OTel.bootstrapTracing(tracerProvider: tracerProvider)

        // --- Logs: bootstrap swift-log to route through OTel ---
        let logExporter = OTLPHTTPLogExporter(
            endpoint: endpointURL.appendingPathComponent("v1/logs"),
            headers: ["Content-Type": "application/x-protobuf"]
        )
        LoggingSystem.bootstrap { label in
            OTelLogHandler(
                label: label,
                exporter: logExporter,
                resource: resource
            )
        }

        // --- Metrics: bootstrap swift-metrics to route through OTel ---
        let metricExporter = OTLPHTTPMetricExporter(
            endpoint: endpointURL.appendingPathComponent("v1/metrics"),
            headers: ["Content-Type": "application/x-protobuf"]
        )
        MetricsSystem.bootstrap(OTelMetricsFactory(
            exporter: metricExporter,
            resource: resource
        ))

        return SequinsHandle(
            tracerProvider: tracerProvider,
            shutdown: {
                try? processor.forceFlush()
            }
        )
    }
}

public enum SequinsError: Error {
    case invalidEndpoint(String)
}

File: sdks/java/src/main/java/dev/sequins/otel/Sequins.java
package dev.sequins.otel;

import io.opentelemetry.api.GlobalOpenTelemetry;
import io.opentelemetry.api.metrics.MeterProvider;
import io.opentelemetry.api.trace.TracerProvider;
import io.opentelemetry.api.logs.LoggerProvider;
import io.opentelemetry.exporter.otlp.logs.OtlpGrpcLogRecordExporter;
import io.opentelemetry.exporter.otlp.metrics.OtlpGrpcMetricExporter;
import io.opentelemetry.exporter.otlp.trace.OtlpGrpcSpanExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.logs.SdkLoggerProvider;
import io.opentelemetry.sdk.logs.export.BatchLogRecordProcessor;
import io.opentelemetry.sdk.metrics.SdkMeterProvider;
import io.opentelemetry.sdk.metrics.export.PeriodicMetricReader;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
import io.opentelemetry.semconv.ServiceAttributes;

/**
 * Sequins OpenTelemetry SDK — zero-config observability for local development.
 *
 * <p>Configures all OTel signals (traces, metrics, logs) to export to a local
 * Sequins instance via OTLP/gRPC. Registers providers globally so
 * auto-instrumentation libraries work without additional wiring.
 *
 * <pre>{@code
 * Sequins sequins = Sequins.init("my-app");
 *
 * Tracer tracer = sequins.getTracerProvider().tracerBuilder("my-module").build();
 * Meter meter = sequins.getMeterProvider().meterBuilder("my-module").build();
 *
 * // On application exit:
 * sequins.close();
 * }</pre>
 */
public final class Sequins implements AutoCloseable {

    private final OpenTelemetrySdk sdk;
    private final SdkTracerProvider sdkTracerProvider;
    private final SdkMeterProvider sdkMeterProvider;
    private final SdkLoggerProvider sdkLoggerProvider;

    private Sequins(
            OpenTelemetrySdk sdk,
            SdkTracerProvider sdkTracerProvider,
            SdkMeterProvider sdkMeterProvider,
            SdkLoggerProvider sdkLoggerProvider) {
        this.sdk = sdk;
        this.sdkTracerProvider = sdkTracerProvider;
        this.sdkMeterProvider = sdkMeterProvider;
        this.sdkLoggerProvider = sdkLoggerProvider;
    }

    /**
     * Initializes OpenTelemetry with all signals pre-configured for Sequins.
     *
     * @param serviceName the name of your service
     * @return a {@link Sequins} handle with references to all configured providers
     */
    public static Sequins init(String serviceName) {
        return init(SequinsConfig.builder().serviceName(serviceName).build());
    }

    /**
     * Initializes OpenTelemetry with all signals pre-configured for Sequins.
     *
     * @param config full configuration; see {@link SequinsConfig#builder()}
     * @return a {@link Sequins} handle with references to all configured providers
     */
    public static Sequins init(SequinsConfig config) {
        String resolvedServiceName = config.getServiceName();
        if (resolvedServiceName == null || resolvedServiceName.isEmpty()) {
            resolvedServiceName = System.getenv("OTEL_SERVICE_NAME");
        }
        if (resolvedServiceName == null || resolvedServiceName.isEmpty()) {
            resolvedServiceName = "unknown_service";
        }

        String endpoint = config.getEndpoint();
        String envEndpoint = System.getenv("OTEL_EXPORTER_OTLP_ENDPOINT");
        if (envEndpoint != null && !envEndpoint.isEmpty()) {
            endpoint = envEndpoint;
        }

        Resource resource = Resource.getDefault().merge(
                Resource.create(io.opentelemetry.api.common.Attributes.of(
                        ServiceAttributes.SERVICE_NAME, resolvedServiceName
                ))
        );

        // --- Traces ---
        OtlpGrpcSpanExporter traceExporter = OtlpGrpcSpanExporter.builder()
                .setEndpoint(endpoint)
                .build();
        SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
                .addSpanProcessor(BatchSpanProcessor.builder(traceExporter).build())
                .setResource(resource)
                .build();

        // --- Metrics ---
        OtlpGrpcMetricExporter metricExporter = OtlpGrpcMetricExporter.builder()
                .setEndpoint(endpoint)
                .build();
        SdkMeterProvider meterProvider = SdkMeterProvider.builder()
                .registerMetricReader(
                        PeriodicMetricReader.builder(metricExporter)
                                .setInterval(config.getMetricExportInterval())
                                .build()
                )
                .setResource(resource)
                .build();

        // --- Logs ---
        OtlpGrpcLogRecordExporter logExporter = OtlpGrpcLogRecordExporter.builder()
                .setEndpoint(endpoint)
                .build();
        SdkLoggerProvider loggerProvider = SdkLoggerProvider.builder()
                .addLogRecordProcessor(BatchLogRecordProcessor.builder(logExporter).build())
                .setResource(resource)
                .build();

        OpenTelemetrySdk sdk = OpenTelemetrySdk.builder()
                .setTracerProvider(tracerProvider)
                .setMeterProvider(meterProvider)
                .setLoggerProvider(loggerProvider)
                .buildAndRegisterGlobal();

        return new Sequins(sdk, tracerProvider, meterProvider, loggerProvider);
    }

    /** Returns the configured {@link TracerProvider}. */
    public TracerProvider getTracerProvider() { return sdkTracerProvider; }

    /** Returns the configured {@link MeterProvider}. */
    public MeterProvider getMeterProvider() { return sdkMeterProvider; }

    /** Returns the configured {@link LoggerProvider}. */
    public LoggerProvider getLoggerProvider() { return sdkLoggerProvider; }

    /**
     * Flushes all pending telemetry and shuts down all providers.
     * Always call this before application exit.
     */
    @Override
    public void close() {
        sdk.close();
    }
}
